cmd/bosun/sched/host.go

```diff
@@ -0,0 +1,809 @@
+						} else {
+							host.Hardware.PowerSupplies[id] = ps
+						}
+					}
+				}
+			}
+		}
+		for _, ts := range hwBatteriesTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
+				slog.Errorf("couldn't find battery id tag %s for host %s", host.Name, id)
+				continue
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.storage.battery", ts.String(), false)
+			status := "Bad"
+			if fstatus == 0 {
+				status = "Ok"
+			}
+			if err == nil && timestamp > 0 {
+				host.Hardware.Storage.Batteries[id] = &Battery{
+					Status:            status,
+					StatusLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwBoardPowerTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.chassis.power.reading", ts.String(), false)
+			if err == nil && timestamp > 0 {
+				host.Hardware.BoardPowerReading = &BoardPowerReading{
+					Watts:            int64(fstatus),
+					WattsLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwPhysicalDiskTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
+				return nil, fmt.Errorf("couldn't find physical disk id tag for host %s", host.Name)
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.storage.pdisk", ts.String(), false)
+			status := "Bad"
+			if fstatus == 0 {
+				status = "Ok"
+			}
+			pd := &PhysicalDisk{}
+			if err == nil && timestamp > 0 {
+				pd.Status = status
+				pd.StatusLastUpdated = timestamp
+				host.Hardware.Storage.PhysicalDisks[id] = pd
+			}
+			for _, m := range hostMetadata {
+				if m.Time.Before(time.Now().Add(-timeFilterAge)) {
+					continue
+				}
+				if !m.Tags.Equal(ts) {
+					continue
+				}
+				switch val := m.Value.(type) {
+				case string:
+					switch m.Name {
+					case "physicalDiskMeta":
+						err = json.Unmarshal([]byte(val), &pd)
+						if err != nil {
+							slog.Errorf("error unmarshalling addresses for host %s, interface %s while generating host api: %s", host.Name, m.Tags["iface"], err)
+						} else {
+							host.Hardware.Storage.PhysicalDisks[id] = pd
+						}
+					}
+				}
+			}
+		}
+		for _, ts := range hwVirtualDiskTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
+				return nil, fmt.Errorf("couldn't find virtual disk id tag for host %s", host.Name)
+			}
+			fstatus, timestamp, err := s.Search.GetLast("hw.storage.vdisk", ts.String(), false)
+			status := "Bad"
+			if fstatus == 0 {
+				status = "Ok"
+			}
+			if err == nil && timestamp > 0 {
+				host.Hardware.Storage.VirtualDisks[id] = &VirtualDisk{
+					Status:            status,
+					StatusLastUpdated: timestamp,
+				}
+			}
+		}
+		for _, ts := range hwControllersTags {
+			if ts["host"] != host.Name {
+				continue
+			}
+			id, ok := ts["id"]
+			if !ok {
```

Comment:
perhaps when making maps we could omit datapoints without required tags. That could eliminate a lot of these checks.